client/allocdir/alloc_dir.go

```diff
@@ -397,3 +430,62 @@ func (d *AllocDir) pathExists(path string) bool {
 	}
 	return true
 }
+
+// WatchSharedDir periodically checks the disk space consumed by the shared
+// allocation directory.
+func (d *AllocDir) StartDiskWatcher() {
+	start := time.Now()
+
+	sync := time.NewTimer(maxCheckDiskInterval)
+	defer sync.Stop()
+
+	d.watchCh = make(chan struct{})
+
+	for {
+		select {
+		case <-d.watchCh:
+			return
+		case <-sync.C:
+			if err := d.syncDiskUsage(); err != nil {
+				log.Printf("[WARN] client: failed to sync disk usage: %v", err)
+			}
+			// calculate the disk ratio
+			diskRatio := float64(d.Size) / float64(d.maxSize*structs.BytesInMegabyte)
+
+			// exponentially decrease the interval when the disk ratio increases
+			nextInterval := time.Duration(int64(1.0/(0.1*math.Pow(diskRatio, 2))+5)) * time.Second
+
+			// Use the maximum interval for the first five minutes or if the
+			// disk ratio is sufficiently high. Also use the minimum check interval
+			// if the disk ratio becomes low enough.
+			if nextInterval < maxCheckDiskInterval || time.Since(start) < checkDiskMaxEnforcePeriod {
+				nextInterval = maxCheckDiskInterval
+			} else if nextInterval > minCheckDiskInterval {
+				nextInterval = minCheckDiskInterval
+			}
+			sync.Reset(nextInterval)
+		}
+	}
+}
+
+// stopDiskWatcher closes the watch channel which causes the disk monitoring to stop.
+func (d *AllocDir) StopDiskWatcher() {
+	close(d.watchCh)
```

Comment:
I would put a `running bool` field and guard against double stops by no-oping `if !d.running`